docs/extensions.rst

```diff
@@ -64,62 +230,121 @@ The following example illustrates a simple extension implementation:
+                opts['project']: {
+                    opts['package']: {
+                        'awesome.py': MY_AWESOME_FILE.format(opts)
+                    },
+                    'tests': {
+                        'awesome_test.py': (
+                            MY_AWESOME_TEST.format(opts),
+                            helpers.NO_OVERWRITE
+                        )
+                    }
                 }
-            }
-        })
-
-        # Files can be directly added to the `structure` dict.
-        scaffold.structure['.python-version'] = ('3.6.1', scaffold.NO_OVERWRITE)
-
-        # The `ensure_file` method can be also used.
-        for filename in opts['awesome_files']:
-            scaffold.ensure_file(filename, content='AWESOME!',
-                                 update_rule=scaffold.NO_CREATE
-                                 path=[opts['project'], 'awesome'])
-
-        # The `reject_file` can be used to avoid files being generated.
-        del scaffold.reject_file('skeleton.py',
-                                 path=[opts['project'], opts['package']])
-
-Note that both :attr:`~pyscaffold.api.Scaffold.before_generate` and
-:attr:`~pyscaffold.api.Scaffold.after_generate` hooks also should be
-defined as a function of a single argument, a
-:class:`~pyscffold.api.Scaffold` instance.
+            })
+
+            structure['.python-version'] = ('3.6.1', helpers.NO_OVERWRITE)
+
+            for filename in opts['awesome_files']:
+                structure = helpers.ensure(
+                    structure, [opts['project'], 'awesome', filename],
+                    content='AWESOME!', update_rule=helpers.NO_CREATE)
+                    # The second argument is the file path, represented by a
+                    # list of file parts or a string.
+                    # Alternatively in this example:
+                    # path = '{project}/awesome/{filename}'.format(
+                    #           filename=filename, **opts)
+
+            # The `reject` can be used to avoid default files being generated.
+            structure = helpers.reject(
+                structure, '{project}/{package}/skeleton.py'.format(**opts))
+                # Alternatively in this example:
+                # path = [opts['project'], opts['package'], 'skeleton.py'])
+
+            # It is import to remember the return values
+            return (structure, opts)
+
+        # ...
+
+.. note::
+
+    The ``project`` and  ``package`` options should be used to provide
+    the correct location of the files relative to the current working
+    directory.
+
+As shown by the previous example, the :mod:`~pyscaffold.api.helpers` argument
+also presents constants that can be used as metadata. The ``NO_OVERWRITE`` flag
+avoids an existing file to be overwritten when ``putup`` is used in update
+mode.  Similarly, ``NO_CREATE`` avoids creating a file from template in update
+mode, even if it does not exist.
+
+.. note::
```

Comment:
Well at that point your remark about `testability` and injection in your mail came back to my mind. Injection the helper module seems to actually complicate things here and as you wrote telling the user to `from pyscaffold import helpers` might be much more natural since not everyone feels comfortable dealing with `functools` or understanding closures in the former case. 
As always it's a tradeoff, is there anything we are winning by doing the injection of `helpers` compared to doing an `from pyscaffold import helpers` inside except of saving the user to write one line? I am not completely sure but in case there is no other reason at least my gut feeling says that the import is less complicated and convoluted.